File: src/main/python/datastructures/trees/trie/trie.py
class Trie:
    class Node:
        def __init__(self):
            self.count = 0
            self.children = [None] * 26

    def __init__(self):
        self.root = self.Node()

    def insert(self, string):
        curr = self.root

        for i in range(len(string)):
            index = ord(string[i]) - ord('a')

            if curr.children[index] is None:
                curr.children[index] = self.Node()

            curr = curr.children[index]

        curr.count = curr.count + 1

    def contains(self, string):
        curr = self.root

        for i in range(len(string)):
            index = ord(string[i]) - ord('a')

            if curr.children[index] is None:
                return False

            curr = curr.children[index]

        return curr.count > 0


if __name__ == "__main__":
    t = Trie()

    t.insert("apple")
    t.insert("app")

    print(t.contains("app"))
    print(t.contains("ap"))

File: java/src/main/java/datastructures/trees/trie/Trie.java
package datastructures.trees.trie;

public class Trie {

  private final TrieNode root;

  public Trie() {
    root = new TrieNode();
  }

  public static void main(String[] args) {
    Trie trie = new Trie();
    trie.insert("app");
    trie.insert("apple");
    trie.insert("ap");

    System.out.println(trie.contains("app"));
    System.out.println(trie.contains("appl"));
  }

  public void insert(String s) {
    TrieNode curr = root;
    char[] chars = s.toCharArray();

    for (char aChar : chars) {
      int index = aChar - 'a';

      if (curr.children[index] == null) {
        curr.children[index] = new TrieNode();
      }

      curr = curr.children[index];
    }

    curr.count++;
  }

  public boolean contains(String s) {
    TrieNode curr = root;
    char[] chars = s.toCharArray();

    for (char aChar : chars) {
      int index = aChar - 'a';

      if (curr.children[index] == null) {
        return false;
      }

      curr = curr.children[index];
    }

    return curr.count > 0;
  }

  private static class TrieNode {
    public final TrieNode[] children;
    public int count;

    public TrieNode() {
      children = new TrieNode[26];
    }
  }
}
